server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1433,4 +1943,16 @@ protected ClusterBlockException checkBlock(UpdateIndexShardSnapshotStatusRequest
             return null;
         }
     }
+
+    private static final class SnapshotFinalization {
+
+        private final SnapshotsInProgress.Entry entry;
+
+        private final Metadata metadata;
```

Comment:
> We would have to write the metadata prior to enqueuing here, right?

Yea that would be one option. I like it in theory, but in practice I'd rather not do that as it makes the whole logic much more complex.

First off, I wonder if this all matters much in practice:

If the snapshot finalizations come in, in a spike then the metadata should be almost the same across all snapshots and with the way we deserialize should be mostly the same objects shouldn't it? I'm having a hard time imagining a situation where finalizations would massively queue up **and** have different metadata. You could have a situation where a snapshot abort or a data node dropping out causes a bunch of concurrent snapshots to fail and get queued up for finalization, but in that case they all come from the same CS update and have the same metadata object. Outside of that, a finalization should generally be as fast or faster than a shard snapshot (I guess there's some strange corner cases with mixed data and master nodes where running shard snapshots make finalization slow for example but even in that case I wouldn't expect those metadata objects to take up a lot of dedicated heap?).


Second of all, I mainly added the metadata to the queue to make things "subjectively more consistent" by reducing the skew between data and metadata snapshot. Technically we don't give any guarantees as to the time of when we take the metadata snapshot (only that it's after all shards have finished).  Maybe it would be enough to simply hold on to the latest metadata seen when movign a snapshot to `SUCESS` state and use that for all the snapshots we finalize. No strong need to too keep a copy per snapshot?

Reply:
> Maybe it would be enough to simply hold on to the latest metadata seen when movign a snapshot to SUCESS state and use that for all the snapshots we finalize. No strong need to too keep a copy per snapshot?

Wouldn't that also be the behavior when we  have a master failover (where the new master now uses a new cluster state as finalization point?

Reply:
Yea exactly ... good point. I'll get rid of this. This "felt" smarter than it is ...